janus_server/src/datastore.rs

```diff
@@ -665,6 +673,90 @@ impl Transaction<'_> {
         })
     }
 
+    // acquire_incomplete_aggregation_jobs retrieves & acquires the IDs of unclaimed incomplete
+    // aggregation jobs. At most `maximum_acquire_count` jobs are acquired. The job is acquired with
+    // a "lease" that will time out; the desired duration of the lease is a parameter, and the lease
+    // expiration time is returned.
+    pub async fn acquire_incomplete_aggregation_jobs(
+        &self,
+        lease_duration: Duration,
+        maximum_acquire_count: usize,
+    ) -> Result<Vec<(TaskId, VdafInstance, AggregationJobId, Time)>, Error> {
+        let now = self.clock.now();
+        let lease_expiry_time = now.add(lease_duration)?;
+        let maximum_acquire_count: i64 = maximum_acquire_count.try_into()?;
+
+        // TODO(brandon): verify that this query is efficient. I am not sure if we would currently
+        // scan over every (in-progress, not-leased) aggregation job for tasks where we are in the
+        // HELPER role.
+        let stmt = self
+            .tx
+            .prepare_cached(
+                "UPDATE aggregation_jobs SET lease_expiry = $1
+                    FROM tasks
+                    WHERE tasks.id = aggregation_jobs.task_id
+                    AND aggregation_jobs.id IN (SELECT aggregation_jobs.id FROM aggregation_jobs
+                        JOIN tasks on tasks.id = aggregation_jobs.task_id
+                        WHERE tasks.aggregator_role = 'LEADER'
+                        AND aggregation_jobs.state = 'IN_PROGRESS'
+                        AND (aggregation_jobs.lease_expiry IS NULL
+                          OR aggregation_jobs.lease_expiry <= $2)
+                        LIMIT $3)
```

Comment:
In this case, the advice in the Postgres documentation falls a little flat, because we don't care much which jobs we get as long as we get some jobs. That is, unpredictability in the subset of rows returned could be seen as a desirable property (since it makes collisions which would lead to serialization failures less likely), rather than a bug.

Still, added LIFO semantics since I think that's what we want in general. I think we might want to eventually rethink what we're LIFO'ing on -- this ordering will make us LIFO on "order of aggregation job creation" when I think we might want to LIFO on "time of last step", but let's discuss this on the bug, this is good enough for now.